source/common/http/filter_manager.h

```diff
@@ -304,6 +306,60 @@ class FilterManagerCallbacks {
    */
   virtual void encodeMetadata(MetadataMapVector& metadata) PURE;
 
+  /**
+   * Injects request trailers into a stream that originally did not have request trailers.
+   */
+  virtual void setRequestTrailers(RequestTrailerMapPtr&& request_trailers) PURE;
+
+  /**
+   * Passes ownership of received continue headers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setContinueHeaders(ResponseHeaderMapPtr&& response_headers) PURE;
+
+  /**
+   * Passes ownership of received response headers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setResponseHeaders(ResponseHeaderMapPtr&& response_headers) PURE;
+
+  /**
+   * Passes ownership of received response trailers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setResponseTrailers(ResponseTrailerMapPtr&& response_trailers) PURE;
+
+  // TODO(snowp): We should consider moving filter access to headers/trailers to happen via the
+  // callbacks instead of via the encode/decode callbacks on the filters.
+
+  /**
+   * The downstream request headers if set.
+   */
+  virtual RequestHeaderMapOptRef requestHeaders() PURE;
+
+  /**
+   * The downstream request trailers if present.
+   */
+  virtual RequestTrailerMapOptRef requestTrailers() PURE;
+
+  /**
+   * Retrieves a pointer to the continue headers set via the call to setContinueHeaders.
+   */
+  virtual ResponseHeaderMapOptRef continueHeaders() PURE;
+
+  /**
+   * Retrieves a pointer to the response headers set via the last call to setResponseHeaders.
+   * Note that response headers might be set multiple times (e.g. if a local reply is issued after
+   * headers have been received but before headers have been encoded), so it is not safe in general
+   * to assume that any set of headers will be valid for the duration of a stream.
+   */
+  virtual ResponseHeaderMapOptRef responseHeaders() PURE;
+
+  /**
+   * Retrieves a pointer to the response trailers set via setResponseTrailers.
```

Comment:
as unlikely as it is to have trailers overwritten it can happen.
via the last call to setResponseTrailers?  here or in a follow-up.

Reply:
Right now there's an ASSERT that verifies that this doesn't happen, but maybe we don't have test coverage for this case?

Reply:
huh, if there were a response filter which held trailers for any length of time that would totally cause problems (though we have no such use case and it's hard to think of a good one so maybe it's fine?)

Reply:
I'll remove the ASSERT and update the comment for now, rather play it safe